### ISSUE TITLE
False Negative: VariableDeclarationUsageDistance

### ISSUE THREAD

**Issue by Vyom-Yadav** [checkstyle/checkstyle#11335]
I have read check documentation: https://checkstyle.sourceforge.io/config_coding.html#VariableDeclarationUsageDistance
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac YOUR_FILE.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance"/>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    public Test test(Object a) {
        int pc = 0; // expected violation
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc).test("abc");
        return this;
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
---

**Describe what you expect in detail.**

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:3:9: Distance between variable 'pc' declaration and its first usage is 7
, but allowed 3.  Consider making that variable final if you still need to store 
its value in advance (before method calls that might have side effects on the original value).
 [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
```
---

Works perfectly fine with following piece of code-
```java
public class Test {
    public Test test(Object a) {
        int pc = 0; // expected violation, got a violation
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc);
        return this;
    }
}
```

**Comment by Vyom-Yadav** [comment]
This was found while working on #11087, I may have the solution at present but cannot say anything with certainty right now, will update as soon the root cause is found.

**Comment by Vyom-Yadav** [comment]
@strkkk False positive?
config for all examples-
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance"/>
    </module>
</module>
```
code-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // ok, all statement in between 
                      // are initialization statements.
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        int k = var + 12;
        return this;
    }
}
```
Output-
```
$ java -jar /home/vyom/Downloads/checkstyle-9.3-all.jar -c config.xml Test.java 
Starting audit...
Audit done.
```
code-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // still ok, all statement in between 
                      // are initialization statements.
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        obj.test("abc");
        ((Test)obj).test("abc");
        int k = var + 12;
        return this;
    }
}
```
output- 
```
$ java -jar /home/vyom/Downloads/checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:4:9: Distance between variable 'var' declaration and its first usage is 7, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
```
No violation was expected as all statements are initialization statements (it could be some setter instead of `test` method), the particular method governing this in check-
https://github.com/checkstyle/checkstyle/blob/eeeb96e26a850133eba894d7ae7282560dd3ba0a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L448-L460

**Comment by strkkk** [comment]
@Vyom-Yadav it looks like a separate issue. I am not very familiar with this check. It should be clearly specified what "initialization" means - invocations of random methods on object instance, or only setters, something else?

**Comment by Vyom-Yadav** [comment]
I couldn't find any set of rules officiating this check, all I could find was
 https://github.com/sevntu-checkstyle/sevntu.checkstyle/issues/32 which adds support for not just setter but for any object calling a method. This was shifted from sevntu to the main repo.

The current implementation only considers initialization statements if they all are from the same object-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // ok, all statement in between 
                      // are initialization statements.
        obj.test("abc"); // --
        obj.test("abc"); //  |
        obj.test("abc"); //  |-----> These all are initialization statements
        obj.test("abc"); //  | 
        obj.test("abc"); //  | 
        obj.test("abc"); // --
        int k = var + 12;
        return this;
    }
}
```
This won't give any error, but if I change it to-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        Test obj2 = test("abc");
        int var = 12; // ok, all statement in between
                      // are initialization statements.
        obj.test("abc"); // --
        obj.test("abc"); //  |
        obj.test("abc"); //  |-----> These all are initialization statements
        obj.test("abc"); //  |       but check shows unexpected behaviour. 
        obj.test("abc"); //  |
        obj2.test("abc");// --
        int k = var + 12;
        return this;
    }
}
```
I will get-
```
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:4:9: Distance between variable 'obj2' declaration and its first usage is 6, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:5:9: Distance between variable 'var' declaration and its first usage is 7, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 2 errors.
```
@nmancus1 @strkkk @romani @rnveach @pbludov Please share your opinion on this check.

**Comment by romani** [comment]
Definition of initializing block is not clear.ain point here is to improve without false positives. So this Check is by design is not very strict as it better to not raise violation on code that should not be changed. Less violation is better than false positive.


But we still can improve our algorithm on this, if we get new idea.

**Comment by Vyom-Yadav** [comment]
@romani These were discovered in the diff report for `FullIdent` in #11231, let's not change any crucial logic because we don't have any solid rule governing this, the simple problem is in this method-
https://github.com/checkstyle/checkstyle/blob/aec751dfba46e7ce6eae351cd93fde43f2cf3105/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L430-L446
This method simply gets instance name of a method call, for eg-
```
obj.fooMethod(arg1);
``` 
this method will simply returns `obj`
but for more complicated method calls like-
```
((SomeCasr)obj).hooMethod(arg1);
```
This method returns `(` as instance name because the current FullIdent is not configured to handle typecast, I modified it in #11231 hence I was able to find this.

**Lets keep it simple and just modify `FullIdent` to handle complicated method calls.**

**Comment by romani** [comment]
We should be accurate with update. There are numerous ways to obfuscate code to do the same but look differently.
If you want to update, only Check should be updated to be a bit more smart.

**Comment by Vyom-Yadav** [comment]
>  If you want to update, only Check should be updated to be a bit more smart.

**I need a set of rules to work on**, smartness can vary from approach to approach, I can't find any rules stating which state what type initialization statements should the check ignore, once we have those rules we can improve the check. Right now the problem is simple-

- `getInstanceName(DetailAST methodCallAst)` returns name of instance which is calling the method
- It is returning **wrong** value for `((SomeCasr)obj).hooMethod(arg1);`
- We will just modify `FullIdent` to handle typcast.

**Comment by romani** [comment]
> We will just modify FullIdent to handle typcast.

no, it is problem of Check, so Check should be updated if we decided to move forward with issue.
`FullIdent` is general class and should be simple and serve simple goal.

-------

I removing approval from issue as 
```
        int pc = 0; // expected violation, but not raised 
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc).test("abc");
        return this;
```

as from user perspective it is better to not break block of `test(...` and keep it same in view as it might easier to see whole logic. Variable scope is very technical nuance that is not a top priority for user to read.

**Comment by nmancus1** [comment]
If we make this a problem of FullIdent, we could impact the following checks:

```
➜  checkstyle git:(master) grep -IR "createFullIdent" | grep -v "Test" | grep java | cut -d ":" -f 1 | uniq
src/main/java/com/puppycrawl/tools/checkstyle/api/FullIdent.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/PackageNameCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/design/VisibilityModifierCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/UncommentedMainCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalCatchCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalInstantiationCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsHashCodeCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PackageDeclarationCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalThrowsCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/CovariantEqualsCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/IllegalImportCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/AvoidStaticImportCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/RedundantImportCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportControlCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/AvoidStarImportCheck.java
src/main/java/com/puppycrawl/tools/checkstyle/utils/AnnotationUtil.java


```

Please show that typecast can be a problem to other checks listed above.

**Comment by Vyom-Yadav** [comment]
> I removing approval from issue as
> 
> ```
>         int pc = 0; // expected violation, but not raised 
>         test("abc").test("abc");
>         test("abc").test("abc");
>         test("abc").test("abc");
>         test("abc").test("abc");
>         test("abc").test("abc");
>         test("abc").test("abc");
>         test(pc).test("abc");
>         return this;
> ```
> 
> as from user perspective it is better to not break block of `test(...` and keep it same in view as it might easier to see whole logic. Variable scope is very technical nuance that is not a top priority for user to read. Keep in mind that Check try to combine technical aspect of visibility with HUMAN ability (not deterministic at all) to read code easily.

@romani I can guarantee that the check does not know about it and is not logically programmed to do so, it is just a bug that is helping that case pass. I will get a violation in this case-
```bash
/var/tmp $ javac YOUR_FILE.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="VariableDeclarationUsageDistance"/>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    public Test test(Object a) {
        int pc = 0; // expected violation
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test("abc").test("abc");
        test(pc);
        return this;
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/TestP/Test.java:3:9: Distance between variable 'pc' declaration and its first usage is 7, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 errors.
```
In this case also user might want to break the block of `test(....)` but they will still get a violation.

I don't get why you can't see that the method is broken-
```java
/**                                                                 
 * Get name of instance whose method is called.                     
 *                                                                  
 * @param methodCallAst                                             
 *        DetailAST of METHOD_CALL.                                 
 * @return name of instance.                                        
 */                                                                 
public static String getInstanceName(DetailAST methodCallAst) { // making it public for testing
    final String methodCallName =                                   
            FullIdent.createFullIdentBelow(methodCallAst).getText();
    final int lastDotIndex = methodCallName.lastIndexOf('.');       
    String instanceName = "";                                       
    if (lastDotIndex != -1) {                                       
        instanceName = methodCallName.substring(0, lastDotIndex);   
    }                                                               
    return instanceName;                                            
}                                                                   
```
Test File-
```java
public class Test {
    public Test test(Object a) {
        a.method(1);
        a.method(1).method2(2);
        return this;
    }
}
```
Test-
```java
@Test                                                                                                       
public void testMethod() throws Exception {                                                                 
    File file = Paths.get(getNonCompilablePath("Test.java")).toFile();                                      
    DetailAST root = JavaParser.parseFile(file, JavaParser.Options.WITHOUT_COMMENTS);                       
                                                                                                            
    // Just navigate to 1st method call                                                                     
    DetailAST methodCallAst1 = root.getFirstChild().getLastChild().getFirstChild().getNextSibling()         
            .getLastChild().getFirstChild().getFirstChild();                                                
    // Get the instance name                                                                                
    String instanceName1 = VariableDeclarationUsageDistanceCheck.getInstanceName(methodCallAst1);           
    assertThat(instanceName1).isEqualTo("a");                                                               
                                                                                                            
    // Just navigate to second method call                                                                  
    DetailAST methodCallAst2 = methodCallAst1.getParent().getNextSibling().getNextSibling().getFirstChild();
    // Get the instance name                                                                                
    String instanceName2 = VariableDeclarationUsageDistanceCheck.getInstanceName(methodCallAst2);           
    assertThat(instanceName2).isEqualTo("(");                                                               
}                                                                                                           
```
This test passes successfully whereas it should not-
The last line **`assertThat(instanceName2).isEqualTo("(");`**, verifies that, instance is having name `(` whereas the actual name is `a`. This is not even typecast, it is failing over a simple chained method call.

**Comment by Vyom-Yadav** [comment]
@nmancus1 Some of these are utils, the complete list is [this](https://vyom-yadav.github.io/DiffReport/diffFullIdentChainedMethodCall/my_check.html) (1 or 2 checks are repeated in the config, won't make a difference).

> Please show that typecast can be a problem to some other checks listed above.

It is a problem for only this check and ChainedMethodCall (currently in development)
https://vyom-yadav.github.io/DiffReport/diffFullIdentChainedMethodCall/index.html

**Comment by nmancus1** [comment]
Also:
```
➜  sevntu.checkstyle git:(master) grep -IR "createFullIdent" | grep -v "Test" | grep java | cut -d ":" -f 1 | uniq
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/MapIterationInForEachLoopCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/AvoidModifiersForTypesCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/ForbidCertainImportsCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/ForbidInstantiationCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/Jsr305AnnotationsCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/EitherLogOrThrowCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/RequireFailForTryCatchInJunitCheck.java
sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/annotation/ForbidAnnotationElementValueCheck.java

```

**Comment by romani** [comment]
ok, you proved me util method has defect.
we can fix it.
But it would mean that we need update(if required) all Checks that use it to make sure that such update does not make false-positives (it might let Checks to do more validation and there might be false positives).
Users do not care about correctness of our util methods, they care that Checks works without false-positives.

Attention: false-negative is ok, at least we do not bother with error in build process in case where we are not sure.

**Comment by Vyom-Yadav** [comment]
> I suspecting that Check should raise violation on code that you provided in issue description, all I say is that provided expectations of Check is NOT ok to my mind, we can discuss it.

Hmm, we have to define rules for what type of method calls should be considered as initialization, currently, the checks work as follows-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12; // ok, all statement in between 
                      // are initialization statements.
        obj.test("abc"); // --
        obj.test("abc"); //  |
        obj.test("abc"); //  |-----> These all are initialization statements
        obj.test("abc"); //  | 
        obj.test("abc"); //  | 
        obj.test("abc"); // --
        int k = var + 12;
        return this;
    }
}
```
No violation was given to the user.

It won't raise a violation only if all the method calls in between are from the same object (`obj` in this case), if I change it to `obj2` somewhere then it will show an error.
Consider another example-
```java
public class Test {
    public Test test(Object a) {
        Test obj = test("abc");
        int var = 12;  // violation, was it expected?
        obj.test("abc"); // --
        obj.test("abc"); //  |
        obj.test("abc"); //  |-----> These all are initialization statements
        obj.test("abc"); //  | 
        obj.test("abc"); //  | 
        obj.test("abc"); // --
        System.out.println(var);
        return this;
    }
}
```
Here we will get a violation-
```
Starting audit...
[ERROR] /home/vyom/IdeaProjects/ConfigCheckstyle/src/test/Test.java:4:9: Distance between variable 'var' declaration and its first usage is 7, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]
Audit done.
Checkstyle ends with 1 error.
```
Now also all method calls start from `obj` except the one that contains `var` i.e. `System.out` so an error is thrown. If no violation was thrown in case 1 then I expect similar results in case 2 or error in both cases.

Forming rules for this is tricky, we would need more inputs to solve this.
You may go through [this](https://vyom-yadav.github.io/DiffReport/diffFullIdentChainedMethodCall/index.html) report to see what changes will be there after fixing just `FullIdent` and not the check itself.

**Comment by romani** [comment]
Ideally violation should be in both cases.
First case should be violation as series of `obj.xxxxx ` does not use `var`.

**Comment by Vyom-Yadav** [comment]
According to me, there shouldn't be a violation in both cases-
The initialization statements in between are not counted, if this was `obj.setVal1()` then these would have made more sense. The documentation right now also says 
`Variable declaration/initialization statements are not counted while calculating length. `

**Comment by Vyom-Yadav** [comment]
I agree we can add this behavior to check to not just only validate method calls associated with instance but to also validate same-named method calls without explicit instance.

**Comment by romani** [comment]
Good analysis, but I highly recommend to make rough implementation and test it in regression execution.

**Comment by Vyom-Yadav** [comment]
@romani Please approve this issue

**Comment by romani** [comment]
I don't agree that case in description is good to have violation. 
In comments there are too much different case discussed, let's. Close this issue and open issue with exact target

**Comment by Vyom-Yadav** [comment]
Closed in favor of #13155
